nebula_hw_interfaces/src/nebula_continental_hw_interfaces/continental_ars548_hw_interface.cpp

```diff
@@ -0,0 +1,739 @@
+
+    sensor_udp_driver_->sender()->open();
+    sensor_udp_driver_->sender()->bind();
+
+    if (!sensor_udp_driver_->sender()->isOpen()) {
+      return Status::ERROR_1;
+    }
+  } catch (const std::exception & ex) {
+    Status status = Status::UDP_CONNECTION_ERROR;
+    std::cerr << status << sensor_configuration_->sensor_ip << ","
+              << sensor_configuration_->data_port << std::endl;
+    return status;
+  }
+  return Status::OK;
+}
+
+Status ContinentalARS548HwInterface::RegisterScanCallback(
+  std::function<void(std::unique_ptr<nebula_msgs::msg::NebulaPackets>)> callback)
+{
+  nebula_packets_reception_callback_ = std::move(callback);
+  return Status::OK;
+}
+
+void ContinentalARS548HwInterface::ReceiveCloudPacketCallbackWithSender(
+  const std::vector<uint8_t> & buffer, const std::string & sender_ip)
+{
+  if (sender_ip == sensor_configuration_->sensor_ip) {
+    ReceiveCloudPacketCallback(buffer);
+  }
+}
+void ContinentalARS548HwInterface::ReceiveCloudPacketCallback(const std::vector<uint8_t> & buffer)
+{
+  if (buffer.size() < sizeof(HeaderPacket)) {
+    PrintError("Unrecognized packet. Too short");
+    return;
+  }
+
+  HeaderPacket header_packet{};
+  std::memcpy(&header_packet, buffer.data(), sizeof(HeaderPacket));
+
+  if (header_packet.service_id.value() != 0) {
+    PrintError("Invalid service id");
+    return;
+  } else if (header_packet.method_id.value() == SENSOR_STATUS_METHOD_ID) {
+    if (
+      buffer.size() != SENSOR_STATUS_UDP_PAYLOAD ||
+      header_packet.length.value() != SENSOR_STATUS_PDU_LENGTH) {
+      PrintError("SensorStatus message with invalid size");
+      return;
+    }
+    ProcessSensorStatusPacket(buffer);
+  } else if (header_packet.method_id.value() == FILTER_STATUS_METHOD_ID) {
+    if (
+      buffer.size() != FILTER_STATUS_UDP_PAYLOAD ||
+      header_packet.length.value() != FILTER_STATUS_PDU_LENGTH) {
+      PrintError("FilterStatus message with invalid size");
+      return;
+    }
+
+    ProcessFilterStatusPacket(buffer);
+  } else if (header_packet.method_id.value() == DETECTION_LIST_METHOD_ID) {
+    if (
+      buffer.size() != DETECTION_LIST_UDP_PAYLOAD ||
+      header_packet.length.value() != DETECTION_LIST_PDU_LENGTH) {
+      PrintError("DetectionList message with invalid size");
+      return;
+    }
+
+    ProcessDataPacket(buffer);
+  } else if (header_packet.method_id.value() == OBJECT_LIST_METHOD_ID) {
+    if (
+      buffer.size() != OBJECT_LIST_UDP_PAYLOAD ||
+      header_packet.length.value() != OBJECT_LIST_PDU_LENGTH) {
+      PrintError("ObjectList message with invalid size");
+      return;
+    }
+
+    ProcessDataPacket(buffer);
+  }
+}
+
+void ContinentalARS548HwInterface::ProcessSensorStatusPacket(const std::vector<uint8_t> & buffer)
+{
+  std::memcpy(&sensor_status_packet_, buffer.data(), sizeof(SensorStatusPacket));
+
+  radar_status_.timestamp_nanoseconds = sensor_status_packet_.stamp.timestamp_nanoseconds.value();
```

Comment:
In `GetRadarStatus()` [continental_ars548_hw_interface.cpp:691](https://github.com/knzo25/nebula/blob/6d3fa3b221d8095518c68f45ce73381d4d7bbd60/nebula_hw_interfaces/src/nebula_continental_hw_interfaces/continental_ars548_hw_interface.cpp#L691) you are using `sensor_status_mutex_` before accessing `radar_status_` but that is the only usage of that mutex. You probably wanted to lock on it in this function as well?

Reply:
Thanks !
Addressed in [2338af7](https://github.com/tier4/nebula/pull/107/commits/2338af727c305890391d6daa356c55319e204dc5)